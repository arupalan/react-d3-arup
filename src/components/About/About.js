import React from 'react';
import about from './README.md';
import Page from '../Page/Page';
import ReactMarkdown from 'react-markdown';

export default class About extends React.Component {
  constructor(props) {
    super(props);
    this.state = { title: 'readme.md', markdown: '' };
  }
  componentDidMount() {
    fetch(about)
      .then(response => {
        return response.text();
      })
      .then(text => {
        this.setState({
          markdown: text,
        });
      });
  }

  render() {
    return (
      <Page {...this.state}>
        <ReactMarkdown source={this.state.markdown} />
      </Page>
    );
  }
}
