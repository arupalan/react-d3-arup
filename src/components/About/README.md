# Frontend Coding Challenge

## Exchange Rates

The following task is to be given to prospective candidates as a skill test.
We will be reviewing both the finished product (UI, UX) as well as the code itself.

## Objective

Build a web app that lets the user enter a date and a base currency and, when submitted,
returns a list of foreign exchange rates from a publicly available API as of that date.

## Requirements

## ● Functional:

1. The app should have the following views (implement whatever navigation
    mechanism you deem suitable):
       ● Homepage
       ● About
       ● Exchange Rates (query and results)
          ● The user must be able to enter the following search parameters:
             ● A Date (i.e. to get the FX rates as of that date)
             ● A Base Currency (e.g. with "GBP" as a base currency, the
                ECB's EUR rate was 1.1745 on the 2​nd​ Jan 2017)

## ● Constraints:

1. Use React and its ecosystem.
2. Do not use Bootstrap or any other pre-defined styling library!
3. If you use a scaffolding tool or template, make sure it is cleaned up and presentable
    before submitting your solution!

## ● General:

1. Write a paragraph or two, explaining the following:
    ● What made you choose the approach you were taking when architecting your
       app?
    ● Explain your design decisions: the layout, the styling options etc. What was
       your thought process?
    ● Did you encounter any problems or challenges? If so, how did you resolve
       them?
2. Apart from the restrictions above, you are free to use any supporting libraries at your
    discretion. Be prepared to justify your choices, however!


### ● Delivery:

1. You can send us the code as a link to a Github or other public repo or as a zipped
    folder
2. Provide quick and simple instructions on how to run the app (ideally be a short
    README in the project).
● **Bonus tasks** ​ (in case the above doesn’t reflect your skill level, add one or both of these
features):
1. Create an additional version of the app, which uses a completely ​ _different_ ​ UI
(different layout and styling, not just different colours!).
2. Add the ability for a user to click on any of the currencies returned in the results table
to display a historical chart of that currency's rate, over the last 12 months.

## Notes

● There is no strict target time for this task, but do give us an indication how long this might
take you.
● One of the simplest APIs for FX rates is ​exchangeratesapi.io​, but feel free to choose another.
● We are primarily looking for good software practices. Do what you feel necessary to deliver a
production-ready solution!
● Let us know if you have any questions or run into any insurmountable problems.
**Good luck!
The Applications Team**
BIG Telematics / ITB Telematics Solutions
