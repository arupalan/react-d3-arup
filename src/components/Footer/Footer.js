
import React from 'react';
import s from './Footer.module.scss';
import Link from '../Link';

class Footer extends React.Component {
  render() {
    return (
      <div className={s.root}>
        <div className={s.container}>
          <span className={s.text}>© React D3 Arup</span>
          <span className={s.spacer}>·</span>
          <Link className={s.link} to="/">
            Home
          </Link>
          <span className={s.spacer}>·</span>
          <Link className={s.link} to="/exchangerates">
            Exchange Rates
          </Link>
          <span className={s.spacer}>·</span>
          <Link className={s.link} to="/privacy">
            Privacy
          </Link>
          <span className={s.spacer}>·</span>
          <Link className={s.link} to="/not-found">
            Not Found
          </Link>
        </div>
      </div>
    );
  }
}

export default Footer;
