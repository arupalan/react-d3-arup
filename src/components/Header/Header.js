import React from 'react';
import s from './Header.module.scss';
import Link from '../Link';
import Navigation from '../Navigation';
import reactd3logo from './react-d3-logo.png';
import reactd3Header from './react-d3-header.png';

class Header extends React.Component {
  render() {
    return (
      <div className={s.root}>
        <div className={s.container}>
          <Navigation />
          <Link className={s.brand} to="/">
            <img
              src={reactd3logo}
              srcSet={`${reactd3logo}`}
              width="38"
              height="38"
              alt="React"
            />
            <span className={s.brandTxt}>
              Data visualization in React using React D3
            </span>
          </Link>
          <div className={s.banner}>
            <img src={reactd3Header} srcSet={`${reactd3Header}`} alt="header" />
            <p className={s.bannerDesc}>
              Complex Data Visualizations made easy
            </p>
          </div>
        </div>
      </div>
    );
  }
}

export default Header;
