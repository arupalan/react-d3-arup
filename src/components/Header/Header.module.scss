@import '../../variables.scss';

:root {
  --brand-color: #61dafb;
}

.root {
  background: #373277;
  color: #fff;
}

.container {
  margin: 0 auto;
  padding: 20px 0;
  max-width: var(--max-content-width);
}

.brand {
  color: lighten(#61dafb, +10%);
  text-decoration: none;
  font-size: 1.75em; /* ~28px */

}

.brandTxt {
  margin-left: 10px;
}

.banner {
  text-align: center;
  img {
    width: 50%;
  }
}

.bannerTitle {
  margin: 0;
  padding: 10px;
  font-weight: normal;
  font-size: 4em;
  line-height: 1em;
}

.bannerDesc {
  padding: 0;
  color: rgba(255, 255, 255, 0.5);
  font-size: 1.25em;
  margin: 0;
}

.pullLeft {
  float: left;
}

.pullRight {
  float: right;
}

.popupHead {
  background-color: #fff;
  clear: both;
  color: #7b7b7b;
  display: inline-table;
  font-size: 21px;
  padding: 7px 10px;
  width: 100%;
  font-family: Oswald;
  & .popupHeadRight {
    margin: 11px 7px 0;
  }
}

.popupHeadLeft img {
  border: 1px solid #7b7b7b;
  border-radius: 50%;
  width: 44px;
}

.popupHeadRight {
  & .btnGroup {
    display: inline-flex;
    margin: 0 8px 0 0;
    vertical-align: top !important;

    & .dropdownMenu {
      border: medium none;
      min-width: 122px;
      padding: 0;
    }

    & .chatHeaderButton {
      background: transparent none repeat scroll 0 0;
      border: 1px solid #636364;
      border-radius: 50%;
      font-size: 14px;
      height: 30px;
      width: 30px;
    }
  }
}
