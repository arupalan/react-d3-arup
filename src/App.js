import React from 'react';
import './App.scss';
import Header from './components/Header/Header';
import Footer from './components/Footer/Footer';
import NotFound from './components/NotFound/NotFound';
import { Router } from '@reach/router';
import About from './components/About/About';
import Chart from './components/Chart';
import ExchangeRates from './components/ExchangeRates';
import { library } from '@fortawesome/fontawesome-svg-core';
import ErrorReport from './components/ErrorReport';
import {
  faEnvelope,
  faKey,
  faChartLine,
  faChartPie,
  faChartArea,
  faSpinner,
  faHome,
} from '@fortawesome/free-solid-svg-icons';

library.add(
  faEnvelope,
  faKey,
  faChartLine,
  faChartPie,
  faChartArea,
  faSpinner,
  faHome
);

function App() {
  return (
    <div>
      <Header />
      <Router>
        <Chart path="/" />
        <About path="/about" />
        <ExchangeRates path="/exchangerates" />
        <Chart path="/chart/:ccypair" />
        <ErrorReport path="/errorReport/:errordetails" />
        <NotFound title="Page Not Found" default />
      </Router>
      <Footer />
    </div>
  );
}

export default App;
