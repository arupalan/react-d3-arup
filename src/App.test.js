import React from 'react';
import App from './App';
import { shallow, mount, configure } from 'enzyme';
import renderer from 'react-test-renderer';
import Header from './components/Header/Header';
import Footer from './components/Footer/Footer';
import About from './components/About/About';
import NotFound from './components/NotFound/NotFound';

describe('>>> App --- Snapshot', () => {
  it('+++capturing Snapshot of App', () => {
    const renderedValue = renderer.create(<App />).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
});

describe('>>> App --- Shallow Renderer REACT COMPONENTS', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should have Header', () => {
    expect(wrapper).toContainReact(<Header />);
  });

  test('should have About', () => {
    expect(wrapper).toContainReact(<About path="/about" />);
  });

  test('should have NotFound', () => {
    expect(wrapper).toContainReact(
      <NotFound title="Page Not Found" default={true} />
    );
  });

  test('should have Footer', () => {
    expect(wrapper).toContainReact(<Footer />);
  });

  test('should not have device or chat', () => {
    expect(wrapper).not.toContainMatchingElements(1, '.device');
    expect(wrapper).not.toContainMatchingElements(1, '.chat');
  });
});

describe('>>> App --- Mount Renderer REACT COMPONENTS', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(<App />);
  });

  test('should have Header', () => {
    expect(wrapper).toContainMatchingElements(1, '.device');
    expect(wrapper).toContainMatchingElements(1, '.chat');
  });
});
