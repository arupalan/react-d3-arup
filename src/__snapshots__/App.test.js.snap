// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> App --- Shallow Renderer REACT COMPONENTS should match snapshot 1`] = `
<div>
  <Header />
  <Router>
    <ChartComponent
      path="/"
    />
    <About
      path="/about"
    />
    <ExchangeRates
      path="/exchangerates"
    />
    <ChartComponent
      path="/chart/:ccypair"
    />
    <ErrorReport
      path="/errorReport/:errordetails"
    />
    <NotFound
      default={true}
      title="Page Not Found"
    />
  </Router>
  <Footer />
</div>
`;

exports[`>>> App --- Snapshot +++capturing Snapshot of App 1`] = `
<div>
  <div
    className="root"
  >
    <div
      className="container"
    >
      <div
        className="root"
        role="navigation"
      >
        <a
          className="link"
          href="/about"
          onClick={[Function]}
        >
          About
        </a>
        <a
          className="link"
          href="/exchangerates"
          onClick={[Function]}
        >
          Exchange Rates
        </a>
        <span
          className="spacer"
        >
           | 
        </span>
        <a
          className="link"
          href="/login"
          onClick={[Function]}
        >
          Log in
        </a>
        <span
          className="spacer"
        >
          or
        </span>
        <a
          className="link highlight"
          href="/register"
          onClick={[Function]}
        >
          Sign up
        </a>
      </div>
      <a
        className="brand"
        href="/"
        onClick={[Function]}
      >
        <img
          alt="React"
          height="38"
          src="react-d3-logo.png"
          srcSet="react-d3-logo.png"
          width="38"
        />
        <span
          className="brandTxt"
        >
          Data visualization in React using React D3
        </span>
      </a>
      <div
        className="banner"
      >
        <img
          alt="header"
          src="react-d3-header.png"
          srcSet="react-d3-header.png"
        />
        <p
          className="bannerDesc"
        >
          Complex Data Visualizations made easy
        </p>
      </div>
    </div>
  </div>
  <div
    role="group"
    style={
      Object {
        "outline": "none",
      }
    }
    tabIndex="-1"
  >
    <div
      className="root"
    >
      <div
        className="container"
      >
        <div
          className="boxHeader"
        >
          <h3
            className="boxTitle pr3"
          >
            MACD D3 Data Visualization
          </h3>
        </div>
        <div
          className="boxBody pr6"
        >
          <div
            className="loader"
          >
            Loading...
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="root"
  >
    <div
      className="container"
    >
      <span
        className="text"
      >
        © React D3 Arup
      </span>
      <span
        className="spacer"
      >
        ·
      </span>
      <a
        className="link"
        href="/"
        onClick={[Function]}
      >
        Home
      </a>
      <span
        className="spacer"
      >
        ·
      </span>
      <a
        className="link"
        href="/exchangerates"
        onClick={[Function]}
      >
        Exchange Rates
      </a>
      <span
        className="spacer"
      >
        ·
      </span>
      <a
        className="link"
        href="/privacy"
        onClick={[Function]}
      >
        Privacy
      </a>
      <span
        className="spacer"
      >
        ·
      </span>
      <a
        className="link"
        href="/not-found"
        onClick={[Function]}
      >
        Not Found
      </a>
    </div>
  </div>
</div>
`;
